### ISSUE TITLE
Conduct market research on c̶a̶t̶a̶l̶o̶g̶s̶ metastores

### ISSUE THREAD

**Issue by astrojuanlu** [kedro-org/kedro-devrel#141]
- June 3rd: Snowflake announces Polaris Catalog, an open source catalog for Apache Iceberg "in the next 90 days" ([announcement](https://www.snowflake.com/blog/introducing-polaris-catalog/))
- June 4th: Databricks announces it has agreed to acquire Tabular ([announcement](https://www.databricks.com/company/newsroom/press-releases/databricks-agrees-acquire-tabular-company-founded-original-creators))
- June 13th: Databricks announces the open sourcing of Unity Catalog https://github.com/unitycatalog/unitycatalog ([announcement](https://www.databricks.com/company/newsroom/press-releases/databricks-open-sources-unity-catalog-creating-industrys-only-open))

At the moment, each platform offers only full read & write capabilities to their own catalog, and read-only capabilities for competitors:

![image](https://github.com/kedro-org/kedro/assets/316517/07a51f39-4dd3-4041-befe-25f986eb6bd3)

([source](https://seattledataguy.substack.com/p/apache-iceberg-what-is-it))

And what's more important: data catalogs aren't new, but we're seeing catalogs created for different use cases and business needs: technical, business, and operational ([source](https://royondata.substack.com/p/the-race-to-own-open-data)).

These are just some open source ones[^1] that have been in the news recently. But there's also [Apache Nessie](https://projectnessie.org/), the Hive Metastore, the Iceberg REST Catalog, probably others I'm missing. Then there are the commercial, vendor-driven ones.

[^1]: counting Polaris as open source

And then we have... the Kedro Catalog!

We've sometimes got questions on "how does the Kedro Catalog compare to the Unity Catalog" - and the answer is that they're complementary, but this is not immediately clear to users (see https://github.com/kedro-org/kedro-plugins/issues/542).

It's very clear that this is going to be a hot topic of discussion in the data engineering space in the coming months so we should have a good answer to how does Kedro interact with all these.

**Comment by datajoely** [comment]
So I think it would be fair to say most mature enterprises will start to align around these table catalogs mostly for governance reasons, but also the other benefits afforded by the open table formats they use.

Confusingly this is all slightly different to a category sometimes called Enterprise Data Catalogs which cover players like [Alation](https://www.alation.com/product/data-catalog/), [Amundsen](https://lfaidata.foundation/tag/amundsen/), [Atlan](https://atlan.com/). These will try and position themselves as a 'catalog of catalogs':
![image](https://github.com/kedro-org/kedro-devrel/assets/35801847/61d8062f-7440-494e-a0c4-21f973388122)

> UC/Polaris open source may change their market position

From a UX point of view I think it would be nice for a user to configure their connections `settings.py` in [how we manage `env` 🤷] and then suddenly all datasets available to those credentials become available in our catalog with minimal boilerplate:

```yaml
my_dataset:
   type: UnityCatalogDataset
   # no credentials or other stuff needed
```

**Comment by astrojuanlu** [comment]
> So I think it would be fair to say most mature enterprises will start to align around these table catalogs mostly for governance reasons, but also the other benefits afforded by the open table formats they use.
> 
> Confusingly this is all slightly different to a category sometimes called Enterprise Data Catalogs which cover players like [Alation](https://www.alation.com/product/data-catalog/), [Amundsen](https://lfaidata.foundation/tag/amundsen/), [Atlan](https://atlan.com/). These will try and position themselves as a 'catalog of catalogs':

Definitely. Someone (or ourselves) need to bring some clarity to the terminology. I really like [this exercise on orchestrators](https://www.run.house/blog/lean-data-automation-a-principal-components-approach) (infrastructure, scheduling, asset) for example.

> From a UX point of view I think it would be nice for a user to configure their connections [...] and then suddenly all datasets available to those credentials become available in our catalog with minimal boilerplate:

We've traditionally mixed the location of the data with the data type (in-memory), which gave us a simple design that however exhibits significant flaws (https://github.com/kedro-org/kedro/issues/1936#issuecomment-1597474696, https://github.com/kedro-org/kedro/issues/770#issuecomment-2115272301). If anything, maybe we should do a better job at separating those, given that your node functions are _intimately tied_ to the in-memory representation of the data (if your node expects a `pd.DataFrame`, it will likely not work with anything else) whereas the transport/location is up for configuration (and could as well be Unity Catalog sometimes, and some other times a blob storage, plain disk storage, an MLflow run, or in-memory).

But this is a topic for https://github.com/kedro-org/kedro/issues/1936.

**Comment by astrojuanlu** [comment]
Update: Unity Catalog now has a UI https://github.com/unitycatalog/unitycatalog-ui

![image](https://github.com/user-attachments/assets/9e649e0d-1888-48ad-9a35-0483d6e0efe5)

There are currently 3 documented integrations (Daft, DuckDB, Trino) http://docs.unitycatalog.io/

And looks like Spark 4.0 + Delta Lake 4.0 might give a better view on how to use Unity Catalog https://books.japila.pl/unity-catalog-internals/spark-integration/#demo

But

- no support for custom S3 endpoints (MinIO) yet https://github.com/unitycatalog/unitycatalog/issues/43
- there are some licensing issues https://github.com/unitycatalog/unitycatalog/issues/75
- and in general the benefits and the usage patterns are still a bit unclear https://github.com/unitycatalog/unitycatalog/issues/231#issuecomment-2232812651

**Comment by astrojuanlu** [comment]
Polaris was open sourced 3 days ago, see https://github.com/polaris-catalog/polaris/pull/2, blog announcement https://www.snowflake.com/blog/polaris-catalog-open-source/

Seems to be based on Apache Nessie by Dremio?

**Comment by astrojuanlu** [comment]
To note, both Unity Catalog and Polaris implement the Apache Iceberg’s REST catalog specification. Looks like we have a winner in terms of HTTP APIs at least.

**Comment by astrojuanlu** [comment]
Some questions I posed about UC https://github.com/unitycatalog/unitycatalog/discussions/208#discussioncomment-10205165

**Comment by astrojuanlu** [comment]
> While other vendor-hosted catalogs deviate from the open source specification, which leads to lock-in, Snowflake’s service for Polaris Catalog is designed to be fully compatible with Polaris Catalog’s open source implementation both now and in the future.

**Comment by astrojuanlu** [comment]
- A good, neutral review https://blog.min.io/catalogs-it-moment/ (a few months old already)
- And I recall @noklam found a LinkedIn comment that explained how Iceberg requires a separate metastore (be it Hive or Glue or else) but I can't locate it anymore

**Comment by astrojuanlu** [comment]
> Recently (~1-2 months) [Daft](https://www.getdaft.io/projects/docs/en/stable/integrations/unity_catalog/) and [Polars](https://docs.pola.rs/api/python/stable/reference/catalog/api/polars.Catalog.scan_table.html) have added options to read Delta tables from a Unity Catalog. If I am reading [this PR from delta-rs](https://github.com/delta-io/delta-rs/pull/3113) correctly, it seems like they are about to add support in python.

_@marrov in https://github.com/kedro-org/kedro-plugins/issues/1013#issue-2859735583_
